opnode/l1/batching.go

```diff
@@ -0,0 +1,132 @@
+package l1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var (
+	TooManyRetries = errors.New("too many retries")
+)
+
+// parallelBatchCall creates a drop-in replacement for the standard batchCallContextFn that splits requests into more batch requests, and will parallelize and retry as configured.
+func parallelBatchCall(log log.Logger, getBatch batchCallContextFn, maxRetry int, maxPerBatch int, maxParallel int) batchCallContextFn {
+	return func(ctx context.Context, requests []rpc.BatchElem) error {
+		return fetchBatched(ctx, log, requests, getBatch, maxRetry, maxPerBatch, maxParallel)
+	}
+}
+
+type batchResult struct {
+	failed  []rpc.BatchElem // if anything has to be retried
+	err     error           // if the batch as a whole failed
+	success int             // amount of items that completed successfully
+}
+
+// fetchBatched fetches the given requests in batches of at most maxPerBatch elements, and with at most maxRetry retries per batch.
+// Batch requests may be split into maxParallel go-routines.
+func fetchBatched(ctx context.Context, log log.Logger, requests []rpc.BatchElem, getBatch batchCallContextFn, maxRetry int, maxPerBatch int, maxParallel int) error {
+	batchRequest := func(ctx context.Context, missing []rpc.BatchElem) (failed []rpc.BatchElem, err error) {
+		if err := getBatch(ctx, missing); err != nil {
+			return nil, fmt.Errorf("failed batch-retrieval of receipts: %v", err)
+		}
+		for _, elem := range missing {
+			if elem.Error != nil {
+				log.Trace("batch request element failed", "err", elem.Error, "elem", elem.Args[0])
+				elem.Error = nil // reset, we'll try this element again
+				failed = append(failed, elem)
+				continue
+			}
+		}
+		return failed, nil
+	}
+
+	// limit capacity, don't write to underlying array on retries
+	requests = requests[:len(requests):len(requests)]
```

Comment:
I prefer an explicit copy or stating that `fetchBatched` modifies `requests`

Reply:
This way we avoid the unnecessary copy of a potentially large request slice in the happy case when no item is retried. This is what the capacity-slicing in golang is meant for: to limit writes to the backing array.